### PR TITLE
[Scalarizer][DirectX] Add support for scalarization of Target  intrinsics

### LINKED ISSUE #107433: Add support for DirectX intrinsics to the scalarizer pass
plan discussed here:
https://github.com/llvm/wg-hlsl/pull/62/files?short_path=ac6e592#diff-ac6e59276afe8016e307eedc5c835f534c0cb353707760b44df0fa9d905a5cf8

### PR DESCRIPTION
**Pull request by farzonl**
Since we are using the Scalarizer pass in the backend we needed a way to allow this pass to operate on Target intrinsics.
We achieved this by adding `TargetTransformInfo ` to the Scalarizer pass. This allowed us to call a function available to the DirectX backend to know if an intrinsic is a target intrinsic that should be scalarized.

### CODE REVIEW

File: llvm/include/llvm/IR/IntrinsicsDirectX.td [hunk 1]
```diff
@@ -72,7 +72,7 @@ def int_dx_udot :
     [llvm_anyint_ty, LLVMScalarOrSameVectorWidth<0, LLVMVectorElementType<0>>],
     [IntrNoMem, Commutative] >;
 
-def int_dx_frac  : DefaultAttrsIntrinsic<[llvm_anyfloat_ty], [LLVMMatchType<0>]>;
+def int_dx_frac  : DefaultAttrsIntrinsic<[llvm_anyfloat_ty], [LLVMMatchType<0>], [IntrNoMem]>;
```
Comment (nikic) [inline llvm/include/llvm/IR/IntrinsicsDirectX.td:75]: These changes should be split out into a separate PR.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp [hunk 2]
```diff
@@ -689,8 +698,10 @@ bool ScalarizerVisitor::splitBinary(Instruction &I, const Splitter &Split) {
   return true;
 }
 
-static bool isTriviallyScalariable(Intrinsic::ID ID) {
-  return isTriviallyVectorizable(ID);
+bool ScalarizerVisitor::isTriviallyScalarizable(Intrinsic::ID ID) {
+
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:702]: Stray newline.

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp [hunk 3]
```diff
@@ -689,8 +698,10 @@ bool ScalarizerVisitor::splitBinary(Instruction &I, const Splitter &Split) {
...
+  return TTI->isTargetIntrinsicTriviallyScalarizable(ID) ||
+         isTriviallyVectorizable(ID);
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/Scalarizer.cpp:704]: Swap these.

Comment (nikic) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:704]: Also, should this check isTargetIntrinsic() first?

File: llvm/test/CodeGen/DirectX/frac.ll [hunk 4]
```diff
@@ -1,31 +1,39 @@
-; RUN: opt -S -dxil-op-lower < %s | FileCheck %s
+; RUN: opt -S -scalarizer -dxil-op-lower -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
```
Comment (nikic) [inline llvm/test/CodeGen/DirectX/frac.ll:1]: Use update_test_checks.py.

**Review by nikic — COMMENTED** [review]
Looks reasonable.

File: llvm/include/llvm/IR/IntrinsicsDirectX.td (hunk 1 above)
Comment (farzonl) [reply llvm/include/llvm/IR/IntrinsicsDirectX.td:75]: https://github.com/llvm/llvm-project/pull/108852

File: llvm/lib/Transforms/Scalar/Scalarizer.cpp (hunk 3 above)
Comment (farzonl) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:704]: When I was first designing this feature I had 
```c++
 bool isTargetIntrinsicTriviallyScalarizable(Intrinsic::ID ID) {
  if (!Function::isTargetIntrinsic(ID))
    return false;
...
}
```
Hence why the names are so similar. However, after I implemented the DirectX part I realized that i'd be returning false anyways if it wasn't a dx intrinsic.

What we could do since I switched the order is to add an assert?

Comment (nikic) [reply llvm/lib/Transforms/Scalar/Scalarizer.cpp:704]: I think this check should be here in Scalarizer, so that isTargetIntrinsicTriviallyScalarizable() doesn't get called with a non-target intrinsic -- it shouldn't be able to influence whether generic intrinsics are scalarizable. So something like

```
if (isTriviallyVectorizable(ID))
  return true;
return Function::isTargetIntrinsic(ID) && TTI->isTargetIntrinsicTriviallyScalarizable(ID);
```
is what I'd do.

File: llvm/test/CodeGen/DirectX/frac.ll (hunk 4 above)
Comment (farzonl) [reply llvm/test/CodeGen/DirectX/frac.ll:2]: done

**Review by llvm-beanz — APPROVED** [review]

**Review by nikic — APPROVED** [review]
LGTM

TTI dependency for Scalarizer should be doubly fine: It's not a canonicalization pass, and the query is limited to target intrinsics only.
